### PR TITLE
proposal for important! options

### LINKED ISSUE #14764: [feature] Recipe definition of options with medium precedence over profiles
### What is your suggestion?

So some recipes can define ``hwloc*:shared=True``, that has higher precedence than profile ``*:shared=False``, but lower than profile ``hwloc*:shared=False``.

cc @uilianries 

### Have you read the CONTRIBUTING guide?

- [ ] I've read the CONTRIBUTING guide

### LINKED ISSUE #15519: [question] Force option values on requirements
### What is your question?

The Documentation of the `configure()` method (https://docs.conan.io/2/reference/conanfile/methods/configure.html) states the following: 

> The definition of options values is only a “suggestion" [...]

Is there any chance to force an option onto a dependency. Or at least automatically check if the set option was actually used?

In our case we have several libraries, all depending on Qt. Some of them need specific modules of Qt, so they configure those in their recipe. But this only seems to have an effect in some cases, totally not obvious when.

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

### PR DESCRIPTION
**Pull request by memsharded**
Changelog: Feature: New "important" options values definition, with higher precedence over regular option value definitions.
Docs: https://github.com/conan-io/docs/pull/3585

Close https://github.com/conan-io/conan/issues/14764
Close https://github.com/conan-io/conan/issues/15519

### CODE REVIEW

**Review by franramirez688 — COMMENTED** [review]
I think the idea is great, but I was wondering if it could be `=!` instead of `!=`. IMO, I think of `!=` could be a bit confusing. About the `strong` name, could it be `override` instead? Would it not be clearer?

**Comment by memsharded** [comment]
>  think the idea is great, but I was wondering if it could be =! instead of !=. IMO, I think of != could be a bit confusing. About the strong name, could it be override instead? Would it not be clearer

The specific syntax is a proposal, but I'd say that the implementation is easier if it is in the option name. Also, in theory, options values can be anything, while options names cannot contain "!" at their definition, which makes a bit more unlikely that someone with an option finishing in ! will make it fail.

Regarding ``override``, I am not sure, it will be ovelapping with the requirements overrides, which is a different concepts. Strong options can also override strong options, and regular options can override regular options... We can look for another name, I agree, but ``override`` I am not sure.

**Comment by AbrilRBS** [comment]
> We can look for another name, I agree, but override I am not sure.

Just by simliarity with CSS's `!important` (Both have an `!`), we can call them important options 🤔

**Comment by czoido** [comment]
> > We can look for another name, I agree, but override I am not sure.
> 
> Just by simliarity with CSS's `!important` (Both have an `!`), we can call them important options 🤔

I like the idea of calling them _important_ options and using the `!`

**Review by czoido — APPROVED** [review]

**Review by AbrilRBS — APPROVED** [review]

File: conans/model/options.py [hunk 1]
```diff
@@ -182,12 +184,19 @@ def __setitem__(self, item, value):
...
+
         current_value = self._data.get(item)
         if self._freeze and current_value.value is not None and current_value != value:
             raise ConanException(f"Incorrect attempt to modify option '{item}' "
                                  f"from '{current_value}' to '{value}'")
         self._ensure_exists(item)
-        self._data.setdefault(item, _PackageOption(item, None)).value = value
+        v = self._data.setdefault(item, _PackageOption(item, None))
+        new_value_important = important or (isinstance(value, _PackageOption) and value.important)
+        if new_value_important or not v.important:
```
Comment (franramirez688) [inline conans/model/options.py:197]: @memsharded Is it not always `v.important == False` in this condition?

Comment (memsharded) [reply conans/model/options.py:197]: Not really, the value could have a previously assigned value, (``setdefault`` will recover that value if it was assigned previously, not create the default one) and if that value was important when it was assigned, the previous ``v.important==True``. This is what avoids previous non important values to be overwritten by downstream values, even if those are not important.

**Review by franramirez688 — APPROVED** [review]

**Pull request merged**